[RA2 Ch6]: Namespace user ID-s gap in Kubernetes (#2252)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter06.md

> **Gap description:** Allocation of hugepages from the same NUMA node as other resources of a Pod. To support this [cAdvisor needed a change to support NUMA](https://github.com/google/cadvisor/pull/2304). Changes in Kubernetes are planned to be implemented in the [Node Topology Manager](https://github.com/kubernetes/enhancements/issues/693).

<a name="6.2.9"></a>
### 6.2.0 User namespaces in Kubernetes

> > **Related requirements:** `e.man.004`, `sec.gen.006`

Reviewer: Why sec.gen.006 - The Platform must support Secure logging. Logging with root account must be prohibited when root privileges are not required.

Suggest sec.sys.007 -- see below for definition

I think instead of just listing the requirement number we should provide the requirement for readability.

```suggestion
> **Related requirements:**
| Reference | Requirement |
|---|---|
| e.man.004 | Capability to isolate resources between tenants |
| sec.sys.007 | The Platform must implement controls enforcing separation of duties and privileges, least privilege use and least common mechanism (Role-Based Access Control). |
```

> **Baseline project:** _Kubernetes v1.20_

> > **Gap description:** Kubernetes does not have a support for user namespaces to provide namespace scoped user ID-s (UID), therefore when an application in a container requires system user privileges the container either needs to run in privileged mode or the infrastructure provides random system UID-s. Scrambling UID-s results in errors when the application needs to set kernel capabilities (e.g.: in case of VLAN trunking) or when a Pod shares data with other Pod-s via persistent storage, therefore none of these solutions are secure or error prone, therefore they are not advised. Support for proper user namespaces in Kubernetes is [under discussion](https://github.com/kubernetes/enhancements/pull/2101).

Reviewer 2: Is "scrambling" the right term? That implies the integrity of the UID itself is messed with, but I think you mean having random or unpredictable UIDs? Not my forte at all so may well be wrong.

Reviewer: also, needs some sentence restructuring as the following is confusing:
"... therefore none of these solutions are secure or error prone ..."
I think what you mean is "... therefore these solutions are error prone or not secure ..." or "... therefore these solutions are either not secure or error prone ..." 

What is "these solutions" referring to? the "privileged mode" and "random system UIDs"?

The sentence starting with "Scrambling" has two consecutive "therefore"? How about the text below?

"Kubernetes does not support namespace scoped user IDs (UIDs). Therefore, when a container-based application in a requires system privileges the container either needs to run in privileged mode or the infrastructure needs to provide random system UIDs. Randomised UIDs result in errors when the application needs to set kernel capabilities (e.g.: in case of VLAN trunking) or when a Pod shares data with other Pods via persistent storage. The "privileged mode" solution is not secure while "random UIDs" is error prone, and therefore these techniques are advised for use. Support for proper user namespaces in Kubernetes is [under discussion](https://github.com/kubernetes/enhancements/pull/2101)."

Author: Tanks @pgoyal01 I use the text you proposed.

<a name="6.2.9"></a>
### 6.2.0 User namespaces in Kubernetes

> **Related requirements:**
| Reference | Requirement |
|---|---|
| e.man.004 | Capability to isolate resources between tenants |
| sec.sys.007 | The Platform must implement controls enforcing separation of duties and privileges, least privilege use and least common mechanism (Role-Based Access Control). |

> **Baseline project:** _Kubernetes v1.20_

> > **Gap description:** Kubernetes does not support namespace scoped user IDs (UIDs). Therefore, when a container-based application in a requires system privileges the container either needs to run in privileged mode or the infrastructure needs to provide random system UIDs. Randomised UIDs result in errors when the application needs to set kernel capabilities (e.g.: in case of VLAN trunking) or when a Pod shares data with other Pods via persistent storage. The "privileged mode" solution is not secure while "random UIDs" is error prone, and therefore these techniques are advised for use. Support for proper user namespaces in Kubernetes is [under discussion](https://github.com/kubernetes/enhancements/pull/2101).

Reviewer: correct typo pointed out by @karinesevilla: " when a container-based application in a requires system privileges " -> 'in a' should be removed; should also remove "in"
typo missing "not" in "... and therefore these techniques are advised for use"; suggest slight modification
```suggestion
> **Gap description:** Kubernetes does not support namespace scoped user IDs (UIDs). Therefore, when a container-based application requires system privileges the container either needs to run in privileged mode or the infrastructure needs to provide random system UIDs. Randomised UIDs result in errors when the application needs to set kernel capabilities (e.g.: in case of VLAN trunking) or when a Pod shares data with other Pods via persistent storage. The "privileged mode" solution is not secure while "random UID" solution is error prone, and therefore these techniques should not be used. Support for proper user namespaces in Kubernetes is [under discussion](https://github.com/kubernetes/enhancements/pull/2101).
```
